### ISSUE TITLE
problem with nn.Mean

### ISSUE THREAD

**Issue by abeschneider** [torch/DEPRECEATED-torch7-distro#32]
The following code causes an error:

``` lua
input = torch.Tensor{1, 2, 3}
seq = nn.Sequential()
seq:add(nn.Mean(1))

seq:forward(input) -- ok
seq:forward(input) -- causes an error on line 10 of Mean.lua
```

Line 10 of Mean.lua has:

``` lua
input.torch.mean(self.output, input, self.dimension)
```

which is fine, but the following line appears to cause the problem:

``` lua
self.output = self.output:select(self.dimension, 1)
```

Thus, when Mean is called again, and if input was a vector,  output will be a scalar the second time around, and will cause torch.mean() to fail. I'm not sure the best fix for this, but in my local copy I added the line at the top of the function:

``` lua
self.output = torch.Tensor()
```

which fixes things for me.

**Comment by clementfarabet** [comment]
Ok, that's definitely an issue. The same problem occurs in nn.Mean(), nn.Min(), nn.Max() and nn.Sum().

The problem is that these modules discard (select) the dimension over which the sum/min/max/mean is done. So the way to use to define your inputs with at least one extra dim. Your example would then be:

``` lua
input = torch.Tensor{1,2,3}:reshape(3,1)
m = nn.Mean(1)
m:forward(input)
m:forward(input)
```

So that fixes your problem, but more generally, I wonder if we should modify these modules or not.

**Comment by abeschneider** [comment]
Hmm.. resizing may not always be easy, depending on where the input is coming from, though, I suppose a Resize module could always be added.

Also, it seems a little non-intuitive to have different behavior for a Tensor of size 3 versus 3x1. I know Numpy has similar behavior, but I've only ever seen it cause problems. Is there a reason to not automatically convert 3 -> 3x1 (or vice versa)?

**Comment by koraykv** [comment]
I was not aware of this issue and I agree that this is not a good behaviour. I changed nn.Sum so that whenever the output is a number it converts it into a tensor. This way all backward compat is maintained....

I think I will change all these modules accordingly. Any objections?

**Comment by clementfarabet** [comment]
No objection, it's more correct to always return tensors...

**Comment by koraykv** [comment]
It is not going to return tensors. If the input is 1D it will still return a number, but the next time updateOutput is called it will check if output is a number and make it into a tensor.
